### PR TITLE
Bugfix systemd docker daemon conflicting with daemon.json

### LINKED ISSUE #22339: Cannot load hosts in configuration json file when starting docker daemon
**Output of `docker version`:**

```
root@10-11-1-156:~# docker version
Client:
 Version:      1.11.0
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   4dc5990
 Built:        Wed Apr 13 18:34:23 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.0
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   4dc5990
 Built:        Wed Apr 13 18:34:23 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
root@10-11-1-156:~# docker info
Containers: 7
 Running: 0
 Paused: 0
 Stopped: 7
Images: 207
Server Version: 1.11.0
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 234
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host
Kernel Version: 3.13.0-83-generic
Operating System: Ubuntu 14.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 1.955 GiB
Name: 10-11-1-156
ID: NDA5:5CVE:Y7Y4:MQJZ:IQVN:ISZ4:5V2Q:JUMU:MGPR:BWGF:LKSR:OW5E
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

two config file, one as flag, one for config loading
```
root@10-11-1-156:~# cat /etc/default/docker
DOCKER_OPTS="-H unix:///var/run/docker.sock"
root@10-11-1-156:~# cat /etc/docker/daemon.json
{"hosts":["tcp://127.0.0.1:2376"]}

```
when execute `service docker restart`:
```
root@10-11-1-156:~# service docker restart
docker stop/waiting
start: Job failed to start
```

check docker daemon's log:
```
Waiting for /var/run/docker.sock
unable to configure the Docker daemon with file /etc/docker/daemon.json: the following directives are specified both as a flag and in the configuration file: hosts: (from flag: [unix:///var/run/docker.sock], from file: [tcp://127.0.0.1:2376])
```

And I was following (https://docs.docker.com/engine/reference/commandline/daemon/#daemon-configuration-file) to add `hosts` in /etc/docker/daemon.json. While fails to start docker daemon.

### PR DESCRIPTION
**Pull request by samrocketman**
**\- What I did**

On Ubuntu package:

```
# apt-cache show docker.io
Package: docker.io
Priority: optional
Section: universe/admin
Installed-Size: 54232
Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Original-Maintainer: Paul Tagliamonte <paultag@debian.org>
Architecture: amd64
Version: 1.12.1-0ubuntu13~16.04.1
Replaces: docker (<< 1.5~)
Depends: adduser, containerd (>= 0.2.3~), iptables, runc (>= 1.0.0~rc1~), init-system-helpers (>= 1.18~), lsb-base (>= 4.1+Debian11ubuntu7), libapparmor1 (>= 2.6~devel), libc6 (>= 2.14), libdevmapper1.02.1 (>= 2:1.02.97)
Recommends: ca-certificates, cgroupfs-mount | cgroup-lite, git, ubuntu-fan, xz-utils, apparmor
Suggests: aufs-tools, btrfs-tools, debootstrap, docker-doc, rinse, zfs-fuse | zfsutils
Breaks: docker (<< 1.5~)
Filename: pool/universe/d/docker.io/docker.io_1.12.1-0ubuntu13~16.04.1_amd64.deb
Size: 10581186
MD5sum: 08676d065c43fc6bcb450cf52c782ed5
SHA1: 8f51b8c7b0fbc3eb879e13a76286cbe13b096d34
SHA256: 0987a665d090d90b0096fbdcab7d18316cad6a1754dbc70f56ab74c36ba53c51
...
```

I created `/etc/default/daemon.json` with the contents.

``` json
{
    "hosts": ["tcp://127.0.0.1:2700"],
    "tlsverify": true,
    "tlscacert": "/etc/docker/ca.pem",
    "tlscert": "/etc/docker/cert.pem",
    "tlskey": "/etc/docker/key.pem"
}
```

Starting the docker daemon (`systemctl start docker.service`) causes the error

> unable to configure the Docker daemon with file
> /etc/docker/daemon.json: the following directives are specified both as
> a flag and in the configuration file: hosts: (from flag: [fd://], from
> file: [tcp://127.0.0.1:2700])

**\- How I did it**

Patch the `docker.service` and the error goes away.

**\- How to verify it**
1. Install the above mentioned `docker.io` package in Ubuntu 16.04.
2. Patch `/lib/systemd/system/docker.service` with the changes in this PR.
3. Restart the daemon and it should succeed.
4. Configure the hosts field of `/etc/docker/daemon.json`.  Restart and watch restart success.

**\- Description for the changelog**

> systemd: fix conflicting hosts option between docker.service and daemon.json

cute cat

fixes #22339

### CODE REVIEW

File: contrib/init/systemd/docker.service [hunk 1]
```diff
@@ -9,7 +9,7 @@ Type=notify
 # the default is not to use systemd for cgroups because the delegate issues still
 # exists and systemd currently does not support the cgroup feature set required
 # for containers run by docker
-ExecStart=/usr/bin/dockerd -H fd://
+ExecStart=/usr/bin/dockerd
```
Comment (thaJeztah) [inline contrib/init/systemd/docker.service]: This change is not correct, as `-H fd://` is required for socket activation (see https://github.com/docker/docker/pull/27473/files#diff-ff907ce70a8c7e795bde1de91be6fa68R5 above)

Comment (samrocketman) [inline contrib/init/systemd/docker.service]: Except, my docker instance works just fine without that option.  That option also directly conflicts with daemon.json.  It's not possible to configure the `hosts` field in `daemon.json` so long as that option exists.

Comment (samrocketman) [inline contrib/init/systemd/docker.service]: FWIW: my `/lib/systemd/system/docker.service` file has...

```
...
ExecStart=/usr/bin/dockerd $DOCKER_OPTS
...
```

**Comment by samrocketman** [comment]
I'm going to open a new issue related to this and amend my commit.

File: contrib/init/systemd/docker.service (hunk 1 above)
Comment (thaJeztah) [inline contrib/init/systemd/docker.service]: FWIW, it's really not recommended to edit the `docker.service` directly (e.g. to add `$DOCKER_OPTS`. Doing so may result in the unit file not being updated if a new version becomes available. To make changes, always use an override ("drop-in") file https://docs.docker.com/engine/admin/systemd/#/custom-docker-daemon-options

Comment (samrocketman) [inline contrib/init/systemd/docker.service]: An excerpt from `/etc/default/docker`:

```
# Here in Debian, this file is sourced by:
#   - /etc/init.d/docker (sysvinit)
#   - /etc/init/docker (upstart)
#   - systemd's docker.service

# Use of this file for configuring your Docker daemon is discouraged.

# The recommended alternative is "/etc/docker/daemon.json", as described in:
#   https://docs.docker.com/v1.11/engine/reference/commandline/daemon/#daemon-configuration-file
```

Notice `Use of this file for configuring your Docker daemon is discouraged`.

Comment (thaJeztah) [inline contrib/init/systemd/docker.service]: That message is (although not incorrect) probably added by the Ubuntu packagers, as it's not in the official packages from this repository https://github.com/docker/docker/blob/v1.12.2/contrib/init/sysvinit-debian/docker.default#L3-L8

The socket is still created with this change, however socket _activation_ no longer works; socket activation allows docker to be automatically started when you're trying to connect to it (and without the service running before that)

Comment (samrocketman) [inline contrib/init/systemd/docker.service]: @thaJeztah what change do you think would be appropriate here?  Do you recognize the issue I'm raising and agree it's an issue?

It seems a bit of a chicken and egg scenario.  I want to use `daemon.json` but `-H fd://` is preventing me (although, I could add that to the hosts of the daemon.json).

**Comment by justincormack** [comment]
We don't recommend using socket activation, it means container restart does
not work for example.

On 18 Oct 2016 5:35 a.m., "Sam Gleske" <notifications@github.com> wrote:

*@samrocketman* commented on this pull request.
------------------------------

In contrib/init/systemd/docker.service
<https://github.com/docker/docker/pull/27473>:

> @@ -9,7 +9,7 @@ Type=notify
 # the default is not to use systemd for cgroups because the delegate
issues still
 # exists and systemd currently does not support the cgroup feature set required
 # for containers run by docker
-ExecStart=/usr/bin/dockerd -H fd://
+ExecStart=/usr/bin/dockerd

Is socket activation necessary here? I stopped the docker daemon and then
ran docker images. It seems it didn't start the docker daemon for me or
perhaps I misunderstand.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/docker/docker/pull/27473>, or mute the thread

**Comment by thaJeztah** [comment]
@justincormack that was for RPM based installs (for which we removed it), on .deb based installs, socket activation is still in use

**Comment by justincormack** [comment]
@thaJeztah why? It causes the same issues. I think we should remove it everywhere.

**Comment by thaJeztah** [comment]
@justincormack I'm not against removing it, as I don't think it's widely used (actually suggested that on other occasions), but based on https://github.com/docker/docker/pull/24804#issuecomment-233693033 its not causing issues for deb-based systems

**Comment by samrocketman** [comment]
> its not causing issues for deb-based systems

It's causing issues on my deb based system because I wish to make use of `daemon.json`.  Have you tried to configure this in any other deb-based system?

**Comment by thaJeztah** [comment]
> It's causing issues on my deb-based system because I wish to make use of daemon.json

@samrocketman that comment was about the reason socket activation was not a problem for deb-based installs for restarting.

A related discussion of the conflict between `daemon.json` and `fd://` / socket activation is here; https://github.com/docker/docker/issues/25471

**Comment by samrocketman** [comment]
Thanks for the reference, I've subscribed for updates.

**Comment by LK4D4** [comment]
@thaJeztah what's up here?

**Comment by thaJeztah** [comment]
@LK4D4 this is about removing socket activation from all systemd unit files, but there was no consensus yet if that's a good thing to do

ping @tonistiigi

**Comment by tonistiigi** [comment]
Socket activation works fine when you just specify it in the json file and leave out the `fd://` from unit file. But I think it would be much more obvious to the user if we would just merge the host arrays instead of forcing the user to use one file or the another.

**Comment by tonistiigi** [comment]
I think socket activation is a useful feature. If we don't use it we are not following the best practices for unit files. It's a shame that selinux and systemd can't figure out how to work together in fedora. For the perspective of specifying hosts, socket activation is no different. We would have the same problem if you specify `-H unix:///var/run/docker.sock` in unit file. The only difference is that we have different rules for merging config and cli options and merging config and default options. If we don't want to merge them I think we could also just have an exception for `fd://`.

**Comment by thaJeztah** [comment]
We discussed this in the maintainers meeting, and this PR in its current form cannot be merged due to it being needed for socket activation.

We are open to having a `config` command that shows the active configuration, and a change in policy for the `daemon.json` to _merge_ list options, instead of error out if both a flag  an `daemon.json` is specified.

I'll open an issue for tracking later

**Comment by nandlalyadav57** [comment]
I am facing an issue while updating below entries in docker daemon.json
{
  "log-driver": "json-file",
  "log-opts": {
    "max-size": "100m",
    "max-file": "5"
  }
}


if i used override options my containers won't come up

[systemm]# cat /lib/systemd/system/docker.service
[Unit]
Description=Docker Application Container Engine
Documentation=http://docs.docker.com
After=network.target
Wants=docker-storage-setup.service
Requires=docker-cleanup.timer

[Service]
Type=notify
NotifyAccess=main
EnvironmentFile=-/run/containers/registries.conf
EnvironmentFile=-/etc/sysconfig/docker
EnvironmentFile=-/etc/sysconfig/docker-storage
EnvironmentFile=-/etc/sysconfig/docker-network
Environment=GOTRACEBACK=crash
Environment=DOCKER_HTTP_HOST_COMPAT=1
Environment=PATH=/usr/libexec/docker:/usr/bin:/usr/sbin
ExecStart=/usr/bin/dockerd-current \
          --add-runtime docker-runc=/usr/libexec/docker/docker-runc-current \
          --default-runtime=docker-runc \
          --exec-opt native.cgroupdriver=systemd \
          --userland-proxy-path=/usr/libexec/docker/docker-proxy-current \
          --init-path=/usr/libexec/docker/docker-init-current \
          --seccomp-profile=/etc/docker/seccomp.json \
          $OPTIONS \
          $DOCKER_STORAGE_OPTIONS \
          $DOCKER_NETWORK_OPTIONS \
          $ADD_REGISTRY \
          $BLOCK_REGISTRY \
          $INSECURE_REGISTRY \
          $REGISTRIES
ExecReload=/bin/kill -s HUP $MAINPID
LimitNOFILE=1048576
LimitNPROC=1048576
LimitCORE=infinity
TimeoutStartSec=0
Restart=on-abnormal
KillMode=process

[Install]
WantedBy=multi-user.target

**Pull request closed without merging**